### PR TITLE
TF-2602 Fix charset is not set when I attach a `text/plain` attachment

### PR DESCRIPTION
**Pull request by dab246**
## Issue

#2602

## Solution

Use [flutter_charset_detector](https://pub.dev/packages/flutter_charset_detector) to detect `charset` of the text file.

## Demo

https://github.com/linagora/tmail-flutter/assets/80730648/ea3daaaa-cf46-44a2-ac93-24c49408d8d2

## Log

Screenshot 2024-02-26 at 13 16 24

## Related

We can review this PR to help it quickly merge

- https://github.com/amake/flutter_charset_detector/pull/6

### CODE REVIEW

**Comment by dab246** [comment]
`Charset` was not successfully set in the returned email body. I have set `charset` for attachment files with `mimeType=text/plain` when sending emails and it was successful. But when opening the received email, in the attachment the `charset` attribute is set by the jmap server to default `us-ascii`. It seems it doesn't accept charset. 

- Payload `Email/set`:

307682918-0070fe97-43c7-4e0a-b86d-ac8c0309a618

- Response `Email/get`

Screenshot 2024-02-29 at 14 12 07

@chibenwa @quantranhong1999

**Comment by dab246** [comment]
`Charset` was not successfully set in the returned email body. I have set `charset` for attachment files with `mimeType=text/plain` when sending emails and it was successful. But when opening the received email, in the attachment the `charset` attribute is set by the jmap server to default `us-ascii`. It seems it doesn't accept charset. 

- Payload `Email/set`:

307682918-0070fe97-43c7-4e0a-b86d-ac8c0309a618

- Response `Email/get`

Screenshot 2024-02-29 at 14 12 07

@chibenwa @quantranhong1999

**Comment by chibenwa** [comment]
What we have in integration tests:

```
{
  "using": ["urn:ietf:params:jmap:core", "urn:ietf:params:jmap:mail"],
  "methodCalls": [
    ["Email/set", {
      "accountId": "$ACCOUNT_ID",
      "create": {
        "aaaaaa": {
          "mailboxIds": {
             "${mailboxId.serialize}": true
          },
          "subject": "World domination",
          "attachments": [
            {
              "blobId": "$blobId",
              "type":"text/plain",
              "charset":"UTF-8",
              "disposition": "attachment",
              "language": ["fr", "en"],
              "location": "http://125.26.23.36/content"
            }
          ]
        }
      }
    }, "c1"],
    ["Email/get",
      {
        "accountId": "$ACCOUNT_ID",
        "ids": ["#aaaaaa"],
        "properties": ["mailboxIds", "subject", "attachments"],
        "bodyProperties": ["partId", "blobId", "size", "name", "type", "charset", "disposition", "cid", "language", "location"]
      },
    "c2"]
  ]
}
```

=> Answer for EMail/get

```
[{
  "mailboxIds": {
    "${mailboxId.serialize}": true
  },
  "subject": "World domination",
  "attachments": [
    {
      "partId": "4",
      "blobId": "${messageId}_4",
      "size": 11,
      "type": "text/plain",
      "charset": "UTF-8",
      "disposition": "attachment",
      "language": ["fr", "en"],
      "location": "http://125.26.23.36/content"
    }
  ]
}]
```

**Comment by chibenwa** [comment]
The contentype shall be specified when uploading the blob:

```
Request method:	POST
Request URI:	http://localhost:34833/upload/29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6
Proxy:			<none>
Request params:	<none>
Query params:	<none>
Form params:	<none>
Path params:	<none>
Headers:		accept=application/json; jmapVersion=rfc-8621
				Content-Type=text/plain; charset=US_ASCII
Cookies:		<none>
Multiparts:		<none>
Body:
[49, 50, 51, 52, 53, 54, 55, 56, 57, 13, 10...
```
```

Likely the Email/set charset should be able to override the one of the upload though...

**Comment by chibenwa** [comment]
https://github.com/apache/james-project/pull/2081

**Comment by dab246** [comment]
- Tested again and it is working well on `preprod` server after BE release.
- We should review and merge it @hoangdat 



https://github.com/linagora/tmail-flutter/assets/80730648/1fafba55-7e59-4dc9-9755-bdba67ca558c

**Comment by tddang-linagora** [comment]
@dab246 Should this PR be rebased to `refactor` or we can review it at current state?

**Review by tddang-linagora — APPROVED** [review]

File: lib/features/upload/data/model/upload_file_arguments.dart [hunk 1]
```diff
@@ -8,13 +9,15 @@ import 'package:tmail_ui_user/features/upload/domain/model/upload_task_id.dart';
 class UploadFileArguments with EquatableMixin {
 
   final DioClient dioClient;
+  final FileUtils fileUtils;
```
Comment (hoangdat) [inline lib/features/upload/data/model/upload_file_arguments.dart:12]: any concerned about isolate?

Comment (dab246) [reply lib/features/upload/data/model/upload_file_arguments.dart:12]: Tested it is not affected, functions in FileUtils access native, storage so not affected by isolate like network

**Comment by dab246** [comment]
> * [x]  @dab246 please rebase and verify Memory leak for this PR. Please check it in both mobile and web.

## Memory leak

No leak

- Web:


https://github.com/user-attachments/assets/57ba6719-3eaa-428f-9851-9f48098a502e


- Mobile:



https://github.com/user-attachments/assets/9b131bc0-a449-4e4b-96e2-9ff61f15b65d

File: lib/features/upload/data/model/upload_file_arguments.dart (hunk 1 above)
Comment (hoangdat) [reply lib/features/upload/data/model/upload_file_arguments.dart:12]: how about failed case, toast also be in weird or not?

Comment (dab246) [reply lib/features/upload/data/model/upload_file_arguments.dart:12]: It has nothing to do with this.

**Pull request merged**